design/src/rfcs/rfc0012_body_callback_apis.md

```diff
@@ -0,0 +1,353 @@
+
+Adding a callback APIs to `ByteStream` and `SdkBody` will enable developers using the SDK to implement things like checksum validations and 'read progress' callbacks.
+
+## The Implementation
+
+*Note that comments starting with '//' are not necessarily going to be included in the actual implementation and are intended as clarifying comments for the purposes of this RFC.*
+
+```rust
+// in aws_smithy_http::callbacks...
+
+// An internal-only type that `SdkBody` interacts with in order to call callbacks
+pub(crate) enum Callback {
+   // A callback to be called when sending requests
+    Send(Box<dyn SendCallback>),
+    // A callback to be called when receiving responses
+    Receive(Box<dyn ReceiveCallback>),
+}
+
+impl Callback {
+   /// This lifecycle function is called for each chunk **successfully** read. If an error occurs while reading a chunk,
+   /// this will not be called. This function takes `&mut self` so that implementors may modify an implementing
+   /// struct/enum's internal state. Implementors may return an error.
+   fn update(&mut self, #[allow(unused_variables)] bytes: &[u8]) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
+      match self {
+         Callback::Send(send_callback) => send_callback.update(bytes),
+         Callback::Receive(receive_callback) => receive_callback.update(bytes),
+      }
+   }
+
+   /// This callback is called once all chunks have been read. If the callback encountered 1 or more errors
+   /// while running `update`s, this is how those errors are raised. Otherwise, this may optionally return
+   /// a [`HeaderMap`][HeaderMap] to be appended to an HTTP body as a trailer or inserted into a request's
+   /// headers.
+   fn finally(
+      &self,
+   ) -> Result<Option<HeaderMap<HeaderValue>>, Box<dyn std::error::Error + Send + Sync>> {
+      match self {
+         Callback::Send(send_callback) => send_callback.headers(),
+         Callback::Receive(receive_callback) => receive_callback.trailers(),
+      }
+   }
+
+   /// Create a new `Callback` from an existing one. This is called when a `Callback` needs to be
+   /// re-initialized with default state. For example: when a request has a body that need to be
+   /// rebuilt, all read callbacks on that body need to be run again but with a fresh internal state.
+   fn make_new(&self) -> Box<dyn BaseCallback> {
+      match self {
+         Callback::Send(send_callback) => send_callback.make_new(),
+         Callback::Receive(receive_callback) => receive_callback.make_new(),
+      }
+   }
+}
+
+/// A callback that, when inserted into a request body, will be called for corresponding lifecycle events.
+// Docs for these methods will mostly be the same as the docs on `Callback` so I've omitted them.
+trait SendCallback: Send {
+   fn update(&mut self, #[allow(unused_variables)] bytes: &[u8]) -> Result<(), BoxError> { Ok(()) }
+   fn headers(
+      &self,
+   ) -> Result<Option<HeaderMap<HeaderValue>>, BoxError> { Ok(None) }
```

Comment:
Hmmm... I'm getting confused by the headers/trailers methods. In the case where we're sending a streaming request to S3, we want to send the checksum as a trailer, right? So is this backwards?

What is the use-case for headers?

Reply:
I'll remove headers